src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,408 @@
+		byte[] dataArray = new byte[data.remaining()];
+		Object payload = dataArray;
+		data.get(dataArray);
+		Message<?> messageToUse = null;
+
+		if (KclMessageDrivenChannelAdapter.this.embeddedHeadersMapper != null) {
+			try {
+				messageToUse = KclMessageDrivenChannelAdapter.this.embeddedHeadersMapper.toMessage((byte[]) payload);
+
+				payload = messageToUse.getPayload();
+			}
+			catch (Exception e) {
+				logger.warn("Could not parse embedded headers. Remain payload untouched.", e);
+			}
+		}
+
+		AbstractIntegrationMessageBuilder<Object> messageBuilder = getMessageBuilderFactory().withPayload(payload)
+				.setHeader(AwsHeaders.RECEIVED_PARTITION_KEY, record.partitionKey())
+				.setHeader(AwsHeaders.RECEIVED_SEQUENCE_NUMBER, record.sequenceNumber());
+
+		if (messageToUse != null) {
+			messageBuilder.copyHeadersIfAbsent(messageToUse.getHeaders());
+		}
+
+		return messageBuilder;
+	}
+
+	private void performSend(AbstractIntegrationMessageBuilder<?> messageBuilder, Object rawRecord) {
+
+		Message<?> messageToSend = messageBuilder.build();
+		try {
+			sendMessage(messageToSend);
+		}
+		catch (Exception e) {
+			logger.error("Got an exception during sending a '" + messageToSend + "'" + "\nfor the '" + rawRecord
+					+ "'.\n" + "Consider to use 'errorChannel' flow for the compensation logic.", e);
+		}
+	}
+
+	public void setStreamInitialSequence(InitialPositionInStream streamInitialSequence) {
+		setStreamInitialSequence(InitialPositionInStreamExtended.newInitialPosition(streamInitialSequence));
+	}
+
+	public void setStreamInitialSequence(InitialPositionInStreamExtended streamInitialSequence) {
+		Assert.notNull(streamInitialSequence, "'streamInitialSequence' must not be null");
+		this.streamInitialSequence = streamInitialSequence;
+	}
+
+	public void setIdleBetweenPolls(int idleBetweenPolls) {
+		this.idleBetweenPolls = Math.max(250, idleBetweenPolls);
+	}
+
+	public void setConsumerBackoff(int consumerBackoff) {
+		this.consumerBackoff = Math.max(1000, consumerBackoff);
+	}
+
+	/**
+	 * Sets the interval between 2 checkpoints.
+	 *
+	 * @param checkpointsInterval interval between 2 checkpoints (in milliseconds)
+	 */
+	public void setCheckpointsInterval(long checkpointsInterval) {
+		this.checkpointsInterval = checkpointsInterval;
+	}
+
+	@Override
+	public String toString() {
+		return "KclMessageDrivenChannelAdapter{consumerGroup='" + this.consumerGroup + '\'' + ", stream='" + this.stream + "'}";
+	}
+
+	private class RecordProcessorFactory implements ShardRecordProcessorFactory {
+		@Override
+		public ShardRecordProcessor shardRecordProcessor() {
+			return new RecordProcessor();
+		}
+	}
+
+	/**
+	 * Processes records and checkpoints progress.
+	 */
+	private class RecordProcessor implements ShardRecordProcessor {
+
+		private String shardId;
+
+		private final Log logger = LogFactory.getLog(RecordProcessor.class);
+
+		// Backoff and retry settings
+		private static final int NUM_RETRIES = 10;
```

Comment:
I don't think we need any retries here. The downstream process is already responsible for proper message handling.
See `RequestHandlerRetryAdvice` for the endpoint to handle message: https://docs.spring.io/spring-integration/docs/current/reference/html/messaging-endpoints-chapter.html#retry-advice